### PR TITLE
Integration Specs

### PR DESCRIPTION
**Pull request by krisbobo**
## Milestone: Fix n+1 problem and integration specs
- Ensured that the N+1 problem is solved when fetching all posts and their comments for a user. 
  - Check out the screenshots below:
 
image

image

- Used Capybara to write integration tests for each view in the project.
 - Tested the following pages:
    - User index page
    - User show page
    - User post index page
    - Post show page

### CODE REVIEW

**Review by BenMukebo — COMMENTED** [review]
## STATUS: INVALID ❌ 

Hi  @krisbobo, @Atril33, 

I can't miss felicitating your efforts and energy, you made a great effort in this project, unfortunately, I cannot proceed to review your code.

## Invalid Code Review Request 

- [ ] You have submitted a project, that doesn't have egal contribution from all members in your pair programming group. Kindly note that this is a pair programming activity and all the members should contributed equally to the project. @Atril33  this is not apply for a README change.
- Note the next reviewer will check the commit history to make sure that everybody has contributed code on this Pull Request.

Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me (@BenMukebo) in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

_Invalid Code Review Request **Does not count** into the code reviews limit.

**Review by afanimartin — CHANGES_REQUESTED** [review]
Hi @krisbobo, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!
<FEEL FREE TO CHANGE THIS MESSAGE TO YOUR PERSONAL WORDS>

## Highlights


## Required Changes ♻️
_Check the comments under the review._
- Unfortunately you haven't fixed the N+1 problem in your Posts controller. N+1 problem typically means you are making more than one request to the database instead one of one.

Screenshot from 2023-07-15 06-46-46


Line 6 of your code should look something like;
```ruby
@posts = @user.posts.includes(:comments)
```
This ensures the Posts are fetched with comments at once. Please make the necessary changes.

- You have failing tests. Please check your code and fix the errors
Screenshot from 2023-07-15 07-01-03

- Your tests lack some required edge cases. Please add the following missing tests
#### User index specs
- Tests that asserts you can see the username of all other users.
- Tests that asserts you can see the profile picture for each user.
- Tests that asserts you can see the number of posts each user has written.
- Tests that redirects to the user's show page when you click on a user.

#### User show specs
* Tests that asserts you can see the user's profile picture.
* Tests that asserts you can see the user's username.
* Tests that asserts you can see the number of posts the user has written.
* Tests that asserts you can see the user's bio.
* Tests that asserts you can see the user's first 3 posts.
* Tests that asserts you can see a button that lets me view all of a user's posts.
* Tests that redirects to the post's show page when you click on a user's post.
* Tests that redirects to the user's post's index page when you click on a **see all posts**.

#### User post index specs
* Tests that asserts you can see the user's profile picture.
* Tests that asserts you can see the user's username.
* Tests that asserts you can see the number of posts the user has written.
* Tests that asserts you can see a post's title.
* Tests that asserts you can see some of the post's body.
* Tests that asserts you can see the first comments on a post.
* Tests that asserts you can see how many comments a post has.
* Tests that asserts you can see how many likes a post has.
* Tests that asserts you can see a section for pagination if there are more posts than fit on the view.
* Tests that redirects to the post's show page when you click on a post.

#### User post show specs
* Tests that asserts you can see the post's title.
* Tests that asserts you can see who wrote the post.
* Tests that asserts you can see how many comments it has.
* Tests that asserts you can see how many likes it has.
* Tests that asserts you can see the post body.
* Tests that asserts you can see the username of each commentor.
* Tests that asserts you can see the comment each commentor left.

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- x
- y
- z

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

## Find me on slack
### Martin Afani

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/features/user_index_spec [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  end
+
+  it 'redirects to the user show page when a user link is clicked' do
+    visit users_path
+
+    first('.user-link').click
+
+    expect(current_path).to eq(user_path(@user1))
+  end
+end
```
Comment (zieeco) [inline spec/features/user_index_spec]: - Hey, team, great job adding the user_index page test,however, the following test below are meant to be added and they're missing. These are required for this project milestone. please, consider adding them:



```
- user show page:
  - I can see the user's profile picture.
  - I can see the user's username.
  - I can see the number of posts the user has written.
  - I can see the user's bio.
  - I can see the user's first 3 posts.
  - I can see a button that lets me view all of a user's posts.
  - When I click a user's post, it redirects me to that post's show page.
  - When I click to see all posts, it redirects me to the user's post's index page.

- User post index page:
  - I can see the user's profile picture.
  - I can see the user's username.
  - I can see the number of posts the user has written.
  - I can see a post's title.
  - I can see some of the post's body.
  - I can see the first comments on a post.
  - I can see how many comments a post has.
  - I can see how many likes a post has.
  - I can see a section for pagination if there are more posts than fit on the view.
  - When I click on a post, it redirects me to that post's show page.

- Post show page:
  - I can see the post's title.
  - I can see who wrote the post.
  - I can see how many comments it has.
  - I can see how many likes it has.
  - I can see the post body.
  - I can see the username of each commentor.
  - I can see the comment each commentor left.
  ```

**Review by zieeco — CHANGES_REQUESTED** [review]
Hi @Team,





Great work on making the changes requested by the previous reviewer 👏 
You've done well implementing some of the changes, however, there are some issues that you still need to work on to go to the next project, but you are almost there!


## To highlight 🎯

- Nice implementing some of the changes.✔️
- Nice design. ✔️
- Descriptive PR. ✔️


## Required Changes ♻️
_Check the comments under the review._




## Optional suggestions

- N/A


**_Every comment with the **[OPTIONAL]** prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were missed by the previous reviewer and addressing them will really improve your application._**


Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by rloterh — COMMENTED** [review]
Hi Christian, 

While you made a great effort in this project, unfortunately, I cannot proceed to review your code.

## Invalid Code Review Request

The requested changes from the previous review have not been implemented. Kindly consider adding the said test scenarios requested in the previous review, those are required.

Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

_Invalid Code Review Request **does not count** into the code reviews limit._
